### ISSUE TITLE
Incompatible with expo-router

### ISSUE THREAD

**Issue by brentvatne** [goatandsheep/react-native-dotenv#501]
**Describe the bug**

Cross-posting from the Expo repository: https://github.com/expo/expo/issues/28933

> So what's happening here is that react-native-dotenv is agressively inlining all environment variables from the machine in the bundle. This has 2 side effects that cause the behavior we are seeing:
> It precedes the babel-preset-expo/src/expo-router-plugin.ts transformation, causing important settings of Expo Router to be overwritten. ([this file](https://github.com/expo/expo/blob/d6515f12fc9eee6d563cc4de32f9aee5c856158f/packages/babel-preset-expo/src/expo-router-plugin.ts#L30))
> In the system environment (outside the bundle), there is an environment variable with the same name used as the require.context within Expo Router. It's from the type generation, and uses the absolute path we are seeing here ([generated here](https://github.com/expo/expo/blob/d6515f12fc9eee6d563cc4de32f9aee5c856158f/packages/%40expo/cli/src/start/server/type-generation/startTypescriptTypeGeneration.ts#L59-L62), set as process.env [here](https://github.com/expo/expo/blob/d6515f12fc9eee6d563cc4de32f9aee5c856158f/packages/%40expo/cli/src/start/server/type-generation/routes.ts#L55))
>
> react-native-dotenv seems to do some internal evaluation (possibly due to static rendering, since we evaluate bundle code). Even when setting safe or allowlist / blocklist, it seems to pollute/mutate all environment variables

**To Reproduce**
1. Run `npx create-expo-app`
2. `cd my-app`
3. `npm install react-native-dotenv`
4. Add the react-native-dotenv babel plugin
5. Boot the app with `npx expo start --ios`
6. Notice that you see "Welcome to Expo" screen rather than the routes in `/app` (without the plugin, you will see tabs and the ` app/(tabs)/index.tsx` screen.

https://github.com/brentvatne/repro-dotenv

**Expected behavior**
react-native-dotenv should play nicely with expo-router.

**Screenshots**
n/a

**Desktop (please complete the following information):**
 - OS: any
 - Browser: any
 - Version: any

**Smartphone (please complete the following information):**
 - Device: any
 - OS: any
 - Browser: any
 - Version: any

**Additional context**
Given this incompatibility, we are currently recommending that developers using Expo Router use [Expo CLI's built-in environment variable support](https://docs.expo.dev/guides/environment-variables/).

I think that one way to fix this issue would be to change the default behavior to whitelisted env vars only. In my opinion, not having this default can be a footgun.

**Comment by byCedric** [comment]
To provide some more context, Expo uses internal environment variables that are not considered public (and therefore should never end up in the bundle). If these variables can be left untouched, we should have a way to get Expo Router and react-native-dotenv to work together.

Basically, all `EXPO_*` variables should be left alone, with the gray area being `EXPO_PUBLIC_*` variables. This is mostly defined by users using the default .env integration in Expo.

**Comment by goatandsheep** [comment]
~~Does it work if you add these to blocklist? I'm happy to add this hint to docs if it works.~~

**Comment by goatandsheep** [comment]
This babel plugin is meant to be simple and non specific. React native dotenv is meant to encourage best practices ~~and you're welcome to go around this using blocklist. Expo-variables is a great alternative, but I'm not sure this is a fault of the library just yet. If you can provide a sample I'd be happy to take a look.~~

**Comment by goatandsheep** [comment]
My bad I didn't see you provided a repro project. I'm taking a look now

**Comment by goatandsheep** [comment]
I also didn't read that blocklist didn't work for you. That is concerning.

**Comment by byCedric** [comment]
Yeah, I'm not exactly sure what's going on. In Expo, we do some bundling and possible static rendering (basically evaluating parts of bundle code in Node). Not sure if that might have anything to do with it. So far, neither `safe` + `path`, `blocklist`, or `allowlist` worked for me, unclear why.

It would be nice if this library can work in Expo, if users need more control over it.

**Comment by danicunhac** [comment]
I'm having the same problem and since my project is still under development I've unshifted `EXPO_PUBLIC_` to my env variables, as specified in https://docs.expo.dev/guides/environment-variables/.

**Comment by mayel15** [comment]
I was facing the same problem. FYI, the incompatibility of `react-native-dotenv` with `expo-router` can load a default screen of `expo`, and not your normal dev app you're building.

![Simulator Screenshot - iPhone SE (3rd generation) - 2024-07-01 at 22 22 23](https://github.com/goatandsheep/react-native-dotenv/assets/97792012/eab3fafe-ed18-4b62-bc73-d4c2d7dbd1b1)

As an alternative for `react-native-dotenv`, they're the default [Expo CLI's built-in environment variable support](https://docs.expo.dev/guides/environment-variables/) or [react-native-config](https://github.com/lugg/react-native-config) for bare react native app.

**Comment by a-eid** [comment]
@mayel15 unfortunately expo's built in env variables is not ideal, rn dotenv is way better.

**Comment by brentvatne** [comment]
@a-eid - can you explain the limitations that you're encountering? you could try disabling them, i'm not sure if that will help here though: https://docs.expo.dev/guides/environment-variables/#disabling-environment-variables

**Comment by a-eid** [comment]
@brentvatne in expo we don't have the ability to load arbitrary .env files. for example `.env.qa` or `.env.staging` you can only load standard .env files `.env.production`, `.env.test` or `.env.development`. 

however with RN dotenv, we're able to do so use the `APP_ENV` env variable.

**Comment by robin-scharf** [comment]
> I was facing the same problem. FYI, the incompatibility of `react-native-dotenv` with `expo-router` can load a default screen of `expo`, and not your normal dev app you're building.
> 
> ![Simulator Screenshot - iPhone SE (3rd generation) - 2024-07-01 at 22 22 23](https://private-user-images.githubusercontent.com/97792012/344982876-eab3fafe-ed18-4b62-bc73-d4c2d7dbd1b1.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjUyNzUwNDcsIm5iZiI6MTcyNTI3NDc0NywicGF0aCI6Ii85Nzc5MjAxMi8zNDQ5ODI4NzYtZWFiM2ZhZmUtZWQxOC00YjYyLWJjNzMtZDRjMmQ3ZGJkMWIxLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MDIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTAyVDEwNTkwN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTRmZGU0MGJkODQ4OTc5MzZiNTJlYmNiMDQ0NmY5ZmVmYzMzYTgwNDg2YWJiYTZhZWVlNDIwNDU1N2FiMjc4MTUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.wUSWCWnWKljb3C_vYwNQ_a3J_YT9rRQIE_2kG6Db4F8)
> 
> As an alternative for `react-native-dotenv`, they're the default [Expo CLI's built-in environment variable support](https://docs.expo.dev/guides/environment-variables/) or [react-native-config](https://github.com/lugg/react-native-config) for bare react native app.

+1 Fortunately, I was able not to depend on react-native-dotenv.

**Comment by cchoi94** [comment]
+1 here, our app is dependent on this package for reasons outlined by @a-eid. Is there something in the works for a fix? 
Would love to update to expo 51 asap.

**Comment by brentvatne** [comment]
@cchoi94 @a-eid - as far as I know there is nothing we need to change on the expo side here, and the issue is in react-native-dotenv as described above by @byCedric. if you'd like to get this library working for you, you could try sending a pr!

**Comment by cchoi94** [comment]
@brentvatne Hey forgot about this issue since I just forked it and made it work on my own. Not well versed in public repo contributions here, so don't know the process. If this code helps anyone, feel free to use, I only changed the index.js. 

https://github.com/cchoi94/react-native-dotenv

Another note, I cloned this repo in efforts to push up a pr, but I'll need access as a contributor @goatandsheep. 

Hope this helps

**Comment by tiavina-mika** [comment]
@cchoi94 's solution works, thanks.
When will it be implemented in the main repository?

**Comment by goatandsheep** [comment]
I am reading this now. Will review and help with merge today

**Comment by goatandsheep** [comment]
Ok i took a look at the code. Will require minor cleanup but overall looks awesome. Will be merging in today.

**Comment by goatandsheep** [comment]
After closer analysis, there are some fundamental concerns that this will cause errors. I will have to spend more time on the weekend figuring out how to fix this

**Comment by samuelpeybernes-dev** [comment]
Is there some news about the fix implementation in the main repository?

**Comment by david-bermejo** [comment]
Hi @goatandsheep, have you had a chance to spend more time on this issue? It's fantastic, and I’d love to continue using it alongside expo-router

**Comment by motdotla** [comment]
Hey everyone,

This is fixed in `v4.1.0` and greater. Thank you for using react-native-dotenv!

cc @brentvatne @david-bermejo @samuelpeybernes-dev @tiavina-mika @mayel15 @byCedric
